pkg/kstatus/observe/reader/cached_reader.go

```diff
@@ -0,0 +1,207 @@
+package reader
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// genGroupKinds captures the generated resources for a given GroupKind. This is used
```

Comment:
Your explanation is good. I have updated the PR to include this. 

The name "generated" resources is a bit unfortunate, because while the most common situation is that an observer for a resource will look at generated resource (like deployment will look at replicasets), it is possible that we want to compute status for a resource by looking other resources that are not generated by that resource. An example from Shopify is that we might compute status for a service by looking if any available pods are targeted by the service. I will update if I come up with a better naming here.